lib/web_ui/dev/steps/compile_bundle_step.dart

```diff
@@ -0,0 +1,291 @@
+  });
+
+  final TestBundle bundle;
+  final bool isVerbose;
+  final Set<FilePath>? testFiles;
+
+  // Maximum number of concurrent compile processes to use.
+  static final int _compileConcurrency = int.parse(io.Platform.environment['FELT_COMPILE_CONCURRENCY'] ?? '8');
+  final Pool compilePool = Pool(_compileConcurrency);
+
+  @override
+  String get description => 'compile_bundle';
+
+  @override
+  bool get isSafeToInterrupt => true;
+
+  @override
+  Future<void> interrupt() async {
+    await cleanup();
+  }
+
+  io.Directory get testSetDirectory => io.Directory(
+    pathlib.join(environment.webUiTestDir.path, bundle.testSet.directory)
+  );
+
+  io.Directory get outputBundleDirectory => getBundleBuildDirectory(bundle);
+
+  List<FilePath> _findTestFiles() {
+    final io.Directory testDirectory = testSetDirectory;
+    if (!testDirectory.existsSync()) {
+      throw ToolExit('Test directory "${testDirectory.path}" for bundle ${bundle.name.ansiMagenta} does not exist.');
+    }
+    return testDirectory
+      .listSync(recursive: true)
+      .whereType<io.File>()
+      .where((io.File f) => f.path.endsWith('_test.dart'))
+      .map<FilePath>((io.File f) => FilePath.fromWebUi(
+          pathlib.relative(f.path, from: environment.webUiRootDir.path)))
+      .toList();
+  }
+
+  TestCompiler _createCompiler() {
+    switch (bundle.compileConfig.compiler) {
+      case Compiler.dart2js:
+        return Dart2JSCompiler(
+          testSetDirectory,
+          outputBundleDirectory,
+          renderer: bundle.compileConfig.renderer,
+          isVerbose: isVerbose,
+        );
+      case Compiler.dart2wasm:
+        return Dart2WasmCompiler(
+          testSetDirectory,
+          outputBundleDirectory,
+          renderer: bundle.compileConfig.renderer,
+          isVerbose: isVerbose,
+        );
+    }
+  }
+
+  @override
+  Future<void> run() async {
+    print('Compiling test bundle ${bundle.name.ansiMagenta}...');
+    final List<FilePath> allTests = _findTestFiles();
+    final TestCompiler compiler = _createCompiler();
+    final Stopwatch stopwatch = Stopwatch()..start();
+    final String testSetDirectoryPath = testSetDirectory.path;
+
+    // Clear out old bundle compilations, if they exist
+    if (outputBundleDirectory.existsSync()) {
+      outputBundleDirectory.deleteSync(recursive: true );
+    }
+
+    final List<Future<MapEntry<String, CompileResult>>> pendingResults = <Future<MapEntry<String, CompileResult>>>[];
+    for (final FilePath testFile in allTests) {
+      final String relativePath = pathlib.relative(
+        testFile.absolute,
+        from: testSetDirectoryPath);
+      final Future<MapEntry<String, CompileResult>> result = compilePool.withResource(() async {
+        if (testFiles != null && !testFiles!.contains(testFile)) {
+          return MapEntry<String, CompileResult>(relativePath, CompileResult.filtered);
+        }
+        final bool success = await compiler.compileTest(testFile);
```

Comment:
The compilation actually writes out a json file that contains the result of the compilation of each test. It still will run any of the tests that actually did compile. I did this because it was really irritating debugging dart2wasm stuff and never being able to actually run any tests at all if a single one failed to compile. However, I need to double check that if some tests fail to compile but all the ones that compile succeed that it actually properly fails and LUCI can pick that up. I'm not positive if I actually did that.

Reply:
It turns out we were immediately failing still, despite my best efforts otherwise 😆 . I went ahead and made a change that allows us to fail compilation and continue, and then bubble up that result at the end.